Add self-hosting guide (#993)
hexpm/hexpm: lib/hexpm_web/templates/docs/self_hosting.html.md

## Self-hosting

> In this guide we'll learn how to serve Hex packages using your own server. First, though, some background on what a Hex server really is.

Reviewer: ```suggestion
In this guide we'll learn how to serve Hex packages on your own package repository. First, though, some background on what a Hex repository really is.
```

### Hex specifications

> A Hex server must follow [Hex specifications](https://github.com/hexpm/specifications). An example server, and a reference implementation, is <https://github.com/hexpm/hexpm> that powers <https://hex.pm>. The Hex team also maintains a lower-level library, <https://github.com/hexpm/hex_core>, that you can use to build and interact with Hex servers.

Reviewer: ```suggestion
A Hex deployment must follow the [Hex specifications](https://github.com/hexpm/specifications). <https://hex.pm> is powered by <https://github.com/hexpm/hexpm> and related services such as <https://github.com/hexpm/hexdocs>. While you can use these projects to run your own Hex infrastructure it is generally not recommended because they include many features and complexity not required by the average deployment. The Hex team also maintains a lower-level library, <https://github.com/hexpm/hex_core>, that you can use to build and interact with Hex services.
```

> A Hex server consists of two [endpoints](https://github.com/hexpm/specifications/blob/master/endpoints.md):

Reviewer: ```suggestion
The specifications describe two [endpoints](https://github.com/hexpm/specifications/blob/master/endpoints.md):
```

1. HTTP API - used for publishing packages, packages search, and administrative tasks.
2. Repository - read-only endpoint that delivers registry resources and package tarballs.

> To run a basic Hex server you only need to implement the Repository endpoint.

Reviewer: ```suggestion
If you only wish to serve packages you only need to implement the Repository endpoint.
```

### Building the registry

> Hex v0.21 introduced a [`mix hex.registry build`](https://hexdocs.pm/hex/Mix.Tasks.Hex.Registry.html) task which provides an easy way to build a local registry.

Reviewer: ```suggestion
Hex v0.21 introduced the [`mix hex.registry build`](https://hexdocs.pm/hex/Mix.Tasks.Hex.Registry.html) task which provides an easy way to build a local registry.
```

If you don't yet have a bucket, create one! By default, files stored on S3 are not publicly accessible. You can enable public access by setting the following bucket policy in your bucket's properties:

```json
{
  "Version": "2008-10-17",
  "Statement": [
    {
      "Sid": "AllowPublicRead",
      "Effect": "Allow",
      "Principal": {
        "AWS": "*"
      },
      "Action": "s3:GetObject",
      "Resource": "arn:aws:s3:::my-bucket/*"
    }
  ]
}
```

You may also consider adding an IAM policy for the user accessing the bucket:

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "VisualEditor0",
      "Effect": "Allow",
      "Action": [
        "s3:PutObject",
        "s3:GetObject",
        "s3:ListBucket",
        "s3:DeleteObject"
      ],
      "Resource": [
        "arn:aws:s3:::my-bucket",
        "arn:aws:s3:::my-bucket/*"
      ]
    }
  ]
}
```

See [Amazon S3 docs](https://docs.aws.amazon.com/s3/index.html) for more information and remember to customize bucket/IAM policies in a way that makes sense for your deployment.

> ### Deploying with Plug.Cowboy & Docker

Reviewer: Maybe this section would be helped with an example github repository that contains the boilerplate and we can trim down on some of the code sections?

Author: Yeah, I was thinking about that too. With the excerpt from mix.exs, two module files, and a full Dockerfile, I thought it's borderline ok to just show all the code instead of linking elsewhere but happy to add a new repo. I think the upside is folks will be able to just clone it but the downside is we'd have yet another thing to maintain and keep in sync with the docs if we ever change things.

Do you have particular code sections you'd like to trim? How? Here's what I was thinking:

I could just show the `children = [...]` part of the supervision tree (I included the full module, because that `require Logger + Logger.info("starting ...")` is really helpful), and trim down the Dockerfile. I thought showing `MyApp.Plug` and `config/runtime.exs` in entirety was important as that's the meat of the project. (It'd be great if there was an easier way to configure Static & BasicAuth plugs at _runtime_ cause that would definitely remove some code wink wink.)

With these small changes in place, would you still go ahead with an external repo? If so, how does hexpm/example_repo sound? I'm happy to put it on my account instead, totally fine either way.

Reviewer: Since you trimmed some of the code I think we can keep it in docs only for now.
